File: Web/ftp-db/frontend/src/actions/report.ts
"use server";

import { OpaqueError, Result } from "@/lib/utils";

export async function reportNote(id: string): Promise<Result<null, string>> {
  const response = await fetch(`${process.env.API_URL}/report`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id,
    }),
  });
  if (!response.ok) {
    const { error } = (await response.json()) as OpaqueError;
    return {
      ok: false,
      error,
    };
  }
  return {
    ok: true,
    result: null,
  };
}

File: Web/ftp-db/backend/src/routes/report.rs
use crate::error::{opaque_error, OpaqueError};
use rocket::http::Status;
use rocket::response::status;
use rocket::serde::json::Json;
use serde::Deserialize;
use std::fmt::Display;
use std::time::Duration;
use thirtyfour::{DesiredCapabilities, WebDriver};

#[derive(Deserialize)]
pub struct ReportRequest {
    id: String,
}

#[post("/report", data = "<body>")]
pub async fn report(body: Json<ReportRequest>) -> Result<(), status::Custom<Json<OpaqueError>>> {
    println!("Request to visit {}", body.id);

    let url = format!("{}/search?id={}", std::env::var("FRONTEND_ADDR").unwrap(), body.id);
    fn default_error<Error: Display>(error: Error) -> status::Custom<Json<OpaqueError>> {
        opaque_error(error, Status::InternalServerError, "Something went wrong")
    }

    let mut caps = DesiredCapabilities::firefox();
    caps.set_headless().map_err(default_error)?;
    let driver = WebDriver::new("http://localhost:4444", caps).await.map_err(default_error)?;
    driver.goto(url).await.map_err(default_error)?;
    tokio::time::sleep(Duration::from_secs(10)).await;

    Ok(())
}